### DISCUSSION TITLE
Arbitrary or Custom Aggregation Function (Jupyter/Pandas/Python)

### DISCUSSION THREAD

**Discussion by eunosm3** [atoti/atoti#896] [Q&A] [answered]
Is it possible to aggregate measures using a non-standard aggregation function?  In my case, I frequently use weighted averages, so the simple `*.MEAN` does not work for me.

**Comment by tibdex** [comment]
Hi, could you share a pseudo code example describing the weighted average you'd like to perform?

**Reply by eunosm3** [reply]
absolutely:
```
def weighted_average(df, val, wgt):

    val = df[val]
    wgt = df[wgt]
    return (val * wgt).sum() / wgt.sum()
```
It's nothing more than multiplying the elements of two columns of data, summing those products, and then dividing that sum by the sum of the "weight" column.

In practice, I use numpy's average function (for reasons that don't matter): 
`np.average(x, weights=df.loc[x.index, weight_col_name])`

From numpy's docs:
"weightsarray_like, optional
An array of weights associated with the values in a. Each value in a contributes to the average according to its associated weight. The array of weights must be the same shape as a if no axis is specified, otherwise the weights must have dimensions and shape consistent with a along the specified axis. If weights=None, then all data in a are assumed to have a weight equal to one. The calculation is:

`avg = sum(a * weights) / sum(weights)`
where the sum is over all included elements. The only constraint on the values of weights is that sum(weights) must not be 0."

**Reply by tibdex** [reply, marked as answer]
You should be able to do that with Atoti too. Let's say that `val` and `wgt` are part of the same `table`. Then you could do:

```python
m["weighted average"] = tt.agg.sum_product(table["val"], table["wgt"]) / tt.agg.sum(table["wgt"])
```

**Reply by eunosm3** [reply]
Perfect!  And so much easier to implement than I expected.  The weighted averages show up right away in the pivot table UI, too.  

Thank you so much for your help.

Atoti is an excellent piece of kit.

**Reply by eunosm3** [reply]
JIC this is helpful to someone who finds this post...

Here's how I added a number of weighted measures:

```python
wgt_calc_flds = ["field1", "field2", "field3", "field4", "field5", "field6"]
wgt_factor = "field_wgt"

for item in wgt_calc_flds:
    m["{}_{}".format(item, "gwm")] = tt.agg.sum_product(
        holdings_data[item],
        holdings_data[wgt_factor],
    ) / tt.agg.sum(holdings_data[wgt_factor])
```

**Reply by eunosm3** [reply]
Also, I did not find this Feb 2022 Atoti article before asking my question.
https://www.atoti.io/features/data-connector/data-connector-pandas/weighted-averages-with-atoti/
